### ISSUE TITLE
Tensorflow does not use the GPU during training with eager execution, despite manual activation

### ISSUE THREAD

**Issue by emadboctorx** [tensorflow/tensorflow#45546]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux (Google colab)):
- TensorFlow installed using `pip install tensorflow-gpu`:
- TensorFlow version (2.3.1):
- Python version: 3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1.243
- GPU model and memory: Tesla T4 computeCapability: 7.5 coreClock: 1.59GHz coreCount: 40 deviceMemorySize: 14.73GiB deviceMemoryBandwidth: 298.08GiB/s

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I'm experiencing a very slow training time when running a third party [script](https://github.com/marload/DeepRL-TensorFlow2/blob/master/DQN/DQN_Discrete.py) which is the same(0% difference between GPU and CPU). I use the following command for activating the GPU on google colab after installing `tensorflow-gpu` using `pip`:

```
physical_devices = tf.config.experimental.list_physical_devices('GPU')
if len(physical_devices) > 0:
    tf.config.experimental.set_memory_growth(physical_devices[0], True)
```
I add the lines above in `main()` in the script I referred to earlier and I use [wandb](https://github.com/wandb/client) for monitoring the training. Here are the [graphs](https://drive.google.com/file/d/1Fgn7tlm6HBUyZIcPelVjxNz_RWvY7QU_/view?usp=sharing) within a few minutes of training showing 0% GPU utilization.

**Describe the expected behavior**

A fast performance which results in a remarkable difference in speeds (CPU vs GPU) and GPU utilization above 0% if the metrics are accurate and if they are not, I'm still experiencing the same speed when run on CPU or GPU.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
import wandb
import tensorflow as tf
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.optimizers import Adam

import gym
import argparse
import numpy as np
from collections import deque
import random

tf.keras.backend.set_floatx('float64')
wandb.init(name='DQN', project="deep-rl-tf2")

parser = argparse.ArgumentParser()
parser.add_argument('--gamma', type=float, default=0.95)
parser.add_argument('--lr', type=float, default=0.005)
parser.add_argument('--batch_size', type=int, default=32)
parser.add_argument('--eps', type=float, default=1.0)
parser.add_argument('--eps_decay', type=float, default=0.995)
parser.add_argument('--eps_min', type=float, default=0.01)

args = parser.parse_args()

class ReplayBuffer:
    def __init__(self, capacity=10000):
        self.buffer = deque(maxlen=capacity)
    
    def put(self, state, action, reward, next_state, done):
        self.buffer.append([state, action, reward, next_state, done])
    
    def sample(self):
        sample = random.sample(self.buffer, args.batch_size)
        states, actions, rewards, next_states, done = map(np.asarray, zip(*sample))
        states = np.array(states).reshape(args.batch_size, -1)
        next_states = np.array(next_states).reshape(args.batch_size, -1)
        return states, actions, rewards, next_states, done
    
    def size(self):
        return len(self.buffer)

class ActionStateModel:
    def __init__(self, state_dim, aciton_dim):
        self.state_dim  = state_dim
        self.action_dim = aciton_dim
        self.epsilon = args.eps
        
        self.model = self.create_model()
    
    def create_model(self):
        model = tf.keras.Sequential([
            Input((self.state_dim,)),
            Dense(32, activation='relu'),
            Dense(16, activation='relu'),
            Dense(self.action_dim)
        ])
        model.compile(loss='mse', optimizer=Adam(args.lr))
        return model
    
    def predict(self, state):
        return self.model.predict(state)
    
    def get_action(self, state):
        state = np.reshape(state, [1, self.state_dim])
        self.epsilon *= args.eps_decay
        self.epsilon = max(self.epsilon, args.eps_min)
        q_value = self.predict(state)[0]
        if np.random.random() < self.epsilon:
            return random.randint(0, self.action_dim-1)
        return np.argmax(q_value)

    def train(self, states, targets):
        self.model.fit(states, targets, epochs=1, verbose=0)
    

class Agent:
    def __init__(self, env):
        self.env = env
        self.state_dim = self.env.observation_space.shape[0]
        self.action_dim = self.env.action_space.n

        self.model = ActionStateModel(self.state_dim, self.action_dim)
        self.target_model = ActionStateModel(self.state_dim, self.action_dim)
        self.target_update()

        self.buffer = ReplayBuffer()

    def target_update(self):
        weights = self.model.model.get_weights()
        self.target_model.model.set_weights(weights)
    
    def replay(self):
        for _ in range(10):
            states, actions, rewards, next_states, done = self.buffer.sample()
            targets = self.target_model.predict(states)
            next_q_values = self.target_model.predict(next_states).max(axis=1)
            targets[range(args.batch_size), actions] = rewards + (1-done) * next_q_values * args.gamma
            self.model.train(states, targets)
    
    def train(self, max_episodes=1000):
        for ep in range(max_episodes):
            done, total_reward = False, 0
            state = self.env.reset()
            while not done:
                action = self.model.get_action(state)
                next_state, reward, done, _ = self.env.step(action)
                self.buffer.put(state, action, reward*0.01, next_state, done)
                total_reward += reward
                state = next_state
            if self.buffer.size() >= args.batch_size:
                self.replay()
            self.target_update()
            print('EP{} EpisodeReward={}'.format(ep, total_reward))
            wandb.log({'Reward': total_reward})


def main():
    physical_devices = tf.config.experimental.list_physical_devices('GPU')
    if len(physical_devices) > 0:
        tf.config.experimental.set_memory_growth(physical_devices[0], True)
    env = gym.make('CartPole-v1')
    agent = Agent(env)
    agent.train(max_episodes=1000)

if __name__ == "__main__":
    main()
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by joehoeller** [comment]
I’m having same issue. But in Kubernetes.

**Comment by emadboctorx** [comment]
@joehoeller the problem is I can't tell whether this is a tensorflow or implementation specific problem. Clearly most of DRL algorithms are designed in a way that is not GPU friendly, a lot of computations are done sequentially using the CPU as this is not the first time that I experience this delay and GPU under-utilization.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2 and TF v2.3. GPU utilization shows 0%. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/e09b444c63660052f3f679a954cdd587/45546.ipynb). 


![gpu 2 3](https://user-images.githubusercontent.com/57165142/101768383-7e9f9a00-3b0b-11eb-91fc-fa51ef78d859.png)
Thanks!

**Comment by emadboctorx** [comment]
There is an update, I'm facing the same issue running the script below I wrote and I noticed that the issue is gone when I disable eager execution, which most probably would be gone for the script I used in the issue description above:

    from collections import deque
    from time import perf_counter
    
    import gym
    import numpy as np
    import tensorflow as tf
    import wandb
    from tensorflow.keras.layers import Conv2D, Dense, Flatten, Input
    from tensorflow.keras.models import Model
    from tensorflow.keras.optimizers import Adam
    
    from utils import activate_gpu_tf
    
    
    class DQN:
        def __init__(self, env, buffer_size=10000, batch_size=32):
            self.env = gym.make(env)
            self.input_shape = self.env.observation_space.shape
            self.main_model = self.create_model()
            self.buffer_size = buffer_size
            self.buffer = deque(maxlen=buffer_size)
            self.batch_size = batch_size
    
        def create_model(self):
            x0 = Input(self.input_shape)
            x = Conv2D(32, 8, 4, activation='relu')(x0)
            x = Conv2D(64, 4, 2, activation='relu')(x)
            x = Conv2D(64, 3, 1, activation='relu')(x)
            x = Flatten()(x)
            x = Dense(512, 'relu')(x)
            x = Dense(self.env.action_space.n)(x)
            return Model(x0, x)
    
        def get_action(self, epsilon, state):
            if np.random.random() < epsilon:
                return self.env.action_space.sample()
            q_values = self.main_model.predict(np.expand_dims(state, 0))
            return np.argmax(q_values)
    
        def get_buffer_sample(self):
            indices = np.random.choice(len(self.buffer), self.batch_size, replace=False)
            memories = [self.buffer[i] for i in indices]
            batch = [np.array(item) for item in zip(*memories)]
            return batch
    
        def update(self, batch, gamma):
            states, actions, rewards, dones, new_states = batch
            q_states = self.main_model.predict(states)
            q_new_states = self.main_model.predict(new_states)
            q_target = np.copy(q_states)
            idx = np.arange(self.batch_size, dtype=np.uint8)
            q_target[idx, actions] = rewards + gamma * np.max(q_new_states, axis=1) * dones
            self.main_model.fit(states, q_target, verbose=0)
    
        def fit(
            self,
            epsilon_start=1,
            epsilon_end=0.01,
            decay_n_frames=150000,
            lr=1e-4,
            gamma=0.99,
            target_reward=19,
        ):
            state = self.env.reset()
            total_rewards = []
            episode_reward = 0
            frame_idx = 0
            start_time = perf_counter()
            optimizer = Adam(lr)
            self.main_model.compile(optimizer, loss='mse')
            mean_reward = 0
            last_reset_frame = 0
            while True:
                frame_idx += 1
                epsilon = max(epsilon_end, epsilon_start - frame_idx / decay_n_frames)
                action = self.get_action(epsilon, state)
                new_state, reward, done, info = self.env.step(action)
                episode_reward += reward
                self.buffer.append((state, action, reward, done, new_state))
                state = new_state
                if done:
                    if mean_reward >= target_reward:
                        print(f'Reward achieved in {frame_idx} frames!')
                        break
                    total_rewards.append(episode_reward)
                    speed = (frame_idx - last_reset_frame) / (perf_counter() - start_time)
                    last_reset_frame = frame_idx
                    start_time = perf_counter()
                    mean_reward = np.mean(total_rewards[-100:])
                    titles = ('frame', 'games', 'reward', 'epsilon', 'speed')
                    values = (
                        frame_idx,
                        len(total_rewards),
                        np.around(mean_reward, 2),
                        np.around(epsilon, 2),
                        f'{round(speed)} frames/s',
                    )
                    display = (f'{title}: {value}' for title, value in zip(titles, values))
                    print(*display, sep=', ')
                    episode_reward = 0
                    state = self.env.reset()
                if len(self.buffer) < self.buffer_size:
                    continue
                batch = self.get_buffer_sample()
                self.update(batch, gamma)
    
    
    if __name__ == '__main__':
        # tf.compat.v1.disable_eager_execution()
        wandb.init(name='Eager execution')
        activate_gpu_tf()
        agn = DQN('PongNoFrameskip-v4', 10000)
        agn.fit()

**Comment by Bennyweasl** [comment]
I am having the same issue when using 2.3.0-gpu-jupyter Docker image. It was updated 13 days ago and we are unable to connect to GPU devices from the Jupyter Notebook. Looking at the logs for the docker container, the dso_loader is looking for the incorrect version of the cuda library (it is trying to load 11 instead of 10.1 per the documentation https://www.tensorflow.org/install/source ) This seems to be a build issue with TensorFlow. I would recommend checking your container logs to see if you are seeing the same issue.   

```
2020-12-22 01:34:00.776696: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-12-22 01:34:00.776935: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcublas.so.11'; dlerror: libcublas.so.11: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-12-22 01:34:00.777110: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcublasLt.so.11'; dlerror: libcublasLt.so.11: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-12-22 01:34:00.777185: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2020-12-22 01:34:00.777235: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2020-12-22 01:34:00.777280: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10
2020-12-22 01:34:00.777408: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcusparse.so.11'; dlerror: libcusparse.so.11: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-12-22 01:34:00.777568: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-12-22 01:34:00.777580: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1757] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
```

**Comment by emadboctorx** [comment]
@Bennyweasl Yes, I remember seeing these "could not load ..." prompts and then I reinstalled cuda based on tensorflow [docs](https://www.tensorflow.org/install/gpu) and the problem persists while on eager execution.

**Comment by Bennyweasl** [comment]
@emadboctorx I went back through my docker cached imaged and saw that when I updated a requirements.txt file. One of the packages forced an update of tensorflow to 2.4.0 which requires cuda 11. That was my issue.

**Comment by emadboctorx** [comment]
@Bennyweasl  yeah, I encountered an issue related to tensorflow 2.4 which is installed by default using pip, being the latest version. I get an error on google colab indicating that there is a problem with cuda driver, I just don't remember the exact error but most probably would be the same as yours.

**Comment by sanjoy** [comment]
> @emadboctorx I went back through my docker cached imaged and saw that when I updated a requirements.txt file. One of the packages forced an update of tensorflow to 2.4.0 which requires cuda 11. That was my issue.

Closing the issue based on this, please reopen if you disagree.

**Comment by Bennyweasl** [comment]
@sanjoy You should not close this issue @emadboctorx was just commenting on something I brought up not his direct issue.

**Comment by emadboctorx** [comment]
@sanjoy I don't remember giving you permission to close my issue.

**Comment by sanjoy** [comment]
Sorry for the confusion.

I was not able to reproduce this though.  I tried both TF 2.3 and latest nightly and Titan-V and P100.  In all cases, I see some memory (400-500 MB) allocated on the GPU.  So clearly the program is doing _something_, but perhaps slowly.

Have you tried using the [TensorBoard profiler](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras) to see if something interesting jumps out?

**Comment by emadboctorx** [comment]
@sanjoy, no problem, I closed the other issue, thanks. I have not tried the profiler no so I will try and see if it shows anything significant.

**Comment by sushreebarsa** [comment]
@emadboctorx Could you please try on latest stable version of tf 2.5 or 2.4.1 and let us know if this is still an issue.Thanks!

**Comment by emadboctorx** [comment]
The problem was solved by modifying the implementation
